### PR TITLE
mischief managed

### PR DESCRIPTION
**Pull request by Hershkovitz-hub**
"Dark times lie ahead of us and there will be a time when we must choose between what is easy and what is right." (Albus Dumbeldore, heading towards python for neuroscience sort-of-hackathon)

### CODE REVIEW

File: hw5.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
+import pathlib
+from pathlib import Path
+from typing import Union, Tuple
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import re
```
Comment (ZviBaratz) [inline hw5.py:7]: Imports can get messy! Help yourself reduce cognitive load by keeping your import statements organized.

File: hw5.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+    """
+
+    def __init__(self, data_fname: Union[pathlib.Path, str]):
+        """Initiate QuestionnaireAnalysis class with following arguments.
+        Arguments:
+            data_fname {Union[pathlib.Path, str]} -- [path to .json file ocntaining subjects' data]
+        """
+        self.data_fname = Path(data_fname)
+        if not self.data_fname.is_file():
+            raise ValueError
```
Comment (ZviBaratz) [inline hw5.py:23]: Try to include a helpful message for your user when you raise an exception, e.g.:
```python
    f"Invalid data file path: {self.data_fname}"
```

File: hw5.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        plt.show()
+        return (hist, bins)
+
+    def check_email(self, email: str):
+        # regex = "^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w)+$"
+        regex = r"[^@]+@[^@]+\.[^@]+"
+        flag = False
+        if re.search(regex, email):
+            flag = True
+        return flag
```
Comment (ZviBaratz) [inline hw5.py:54]: Consider:
```python
    return re.match(regex, email)
```

File: hw5.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+            -------
+            df : pd.DataFrame
+            A corrected DataFrame, i.e. the same table but with the erroneous rows removed and
+            the (ordinal) index after a reset.
+        """
+        self.read_data()
+        df = self.data.copy()
+        for i in df.index:
+            if not self.check_email(df.email[i]):
+                df.drop(i, inplace=True)
```
Comment (ZviBaratz) [inline hw5.py:69]: In the future, consider using `df.apply()` for this kind of computation.

File: hw5.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    def fill_na_with_mean(self) -> Tuple[pd.DataFrame, np.ndarray]:
+        """Finds, in the original DataFrame, the subjects that didn't answer
+        all questions, and replaces that missing value with the mean of the
+        other grades for that student.
+
+    Returns
+    -------
+    df : pd.DataFrame
+    The corrected DataFrame after insertion of the mean grade
+    arr : np.ndarray
```
Comment (ZviBaratz) [inline hw5.py:82]: Mind your indentation!

**Comment by ZviBaratz** [comment]
> "Dark times lie ahead of us and there will be a time when we must choose between what is easy and what is right." (Albus Dumbeldore, heading towards python for neuroscience sort-of-hackathon)

It's spelled Dumb**le**dore.
_-10 pts._

**Comment by Hershkovitz-hub** [comment]
By Merlin's beard, how did I make such a mistake?
 Thank you for having mercy and only substracting 10 points.
Gal Hershkovitz
PhD candidate

Sagol school of neuroscience,
Tel-Aviv university
m: +972-509024295
e: hershkovitz1@mail.tau.ac.il
[image: Banner] [image: Facebook icon][image: Instagram icon]

**Pull request closed without merging**